hathor/indexes/memory_deps_index.py

```diff
@@ -0,0 +1,170 @@
+            # XXX: this check serves to avoid calling is_ready() when nothing changed
+            if deps[tx] != validation:
+                deps[tx] = validation
+                if deps.is_ready():
+                    self._del_from_deps_index(cousin)
+                    self._txs_with_deps_ready.add(cousin)
+
+    def update(self, tx: BaseTransaction, partial: bool = True) -> None:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        validation = get_validation_state(tx.storage, tx.hash)
+        if validation.is_fully_connected():
+            self._del_from_deps_index(tx.hash)
+        elif not partial:
+            raise ValueError('partial=False will only accept fully connected transactions')
+        else:
+            self._add_deps(tx)
+            self._add_needed(tx)
+
+    def _add_deps(self, tx: BaseTransaction) -> None:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        # deps are immutable for a given hash
+        _deps = _DirDepValue((dep, get_validation_state(tx.storage, dep)) for dep in tx.get_all_dependencies())
+        # short circuit add directly to ready
+        if _deps.is_ready():
+            self._txs_with_deps_ready.add(tx.hash)
+            return
+        # add direct deps
+        if __debug__ and tx.hash in self._dir_dep_index:
+            # XXX: dependencies set must be immutable
+            assert self._dir_dep_index[tx.hash].keys() == _deps.keys()
+        self._dir_dep_index[tx.hash] = _deps
+        # add reverse dep
+        for rev_dep in _deps:
+            if rev_dep not in self._rev_dep_index:
+                self._rev_dep_index[rev_dep] = set()
+            self._rev_dep_index[rev_dep].add(tx.hash)
+
+    def _del_from_deps_index(self, tx: bytes) -> None:
+        _deps = self._dir_dep_index.pop(tx, _DirDepValue())
+        for rev_dep in _deps.keys():
+            rev_deps = self._rev_dep_index[rev_dep]
+            if tx in rev_deps:
+                rev_deps.remove(tx)
+            if not rev_deps:
+                del self._rev_dep_index[rev_dep]
+
+    def remove_ready_for_validation(self, tx: bytes) -> None:
+        """ Removes from ready for validation set.
+        """
+        self._txs_with_deps_ready.discard(tx)
+
+    def next_ready_for_validation(self, tx_storage: 'TransactionStorage', *, dry_run: bool = False) -> Iterator[bytes]:
+        if dry_run:
+            cur_ready = self._txs_with_deps_ready.copy()
+        else:
+            cur_ready, self._txs_with_deps_ready = self._txs_with_deps_ready, set()
+        while cur_ready:
+            yield from sorted(cur_ready, key=lambda tx_hash: tx_storage.get_transaction(tx_hash).timestamp)
+            if dry_run:
+                cur_ready = self._txs_with_deps_ready - cur_ready
+            else:
+                cur_ready, self._txs_with_deps_ready = self._txs_with_deps_ready, set()
+
+    def iter(self) -> Iterator[bytes]:
```

Comment:
Why not `iter_rev_dep()`?

Reply:
That seems just as confusing. I think the docstring's explanation isn't great, I'll improve it. But basically this is almost like the "main", most obvious iterator, although it is only used in one place (during initialization), but the point of this iterator is to yield any transaction that is not fully validated in the index, so that we can check if we can step the validation.

This is likely redundant, because if everything is consistent within the index, downloading the "needed txs" and stepping through the "next to validate", would be enough to step through all the transactions.

Reply:
Makes sense. Thanks!